feat: IXamlMethod Generic definition

workgroupengineering opened the pull request:
Closes #112

workgroupengineering committed:
feat(IXamlMethod): Add IsGenericMethodDefinition, GenericParameters, and GenericArguments

workgroupengineering committed:
test

workgroupengineering committed:
fix: IXamlType Name Inconsistency from Sre e Cecil when is IXamlType is GenericParameter

workgroupengineering committed:
fix

workgroupengineering committed:
missig

kekekeks commented on tests/TypeSystemTest/BaseTest.cs:
```diff
@@ -0,0 +1,197 @@
+    [Fact]
+    public void Generic_Types_Are_Correctly_Handled()
+    {
+        var wut = TypeSystem.GetType("TypeSystemTest.Models.TestType");
```
As much as I like short variable names, they make the code way less readable when coming in such abundance. Can we have a bit more descriptive ones?

kekekeks commented on src/XamlX.IL.Cecil/CecilMethod.cs:
```diff
@@ -82,6 +82,27 @@ public CecilMethodBase(CecilTypeResolveContext typeResolveContext, MethodReferen
+            public bool IsGenericMethod => Reference.HasGenericParameters;
+            public bool IsGenericMethodDefinition => Reference.IsDefinition && Reference.HasGenericParameters;
+
+            public IReadOnlyList<IXamlType> GenericParameters => _genericParameters ??=
```
I wonder how well would it handle scenarios like:
```cs
class Foo<TFoo>
{
    class Bar<TBar>
    {
         public void Test<TMethod>(TFoo foo, TBar bar, TMethod method) {}
    }
}
```

workgroupengineering replied:
Currently neither the Cecil nor Sre TypeSystem support nested classes.   For Cecil there is this PR #54 to enable support.

In Sre, I removed the following line for a quick test

https://github.com/kekekeks/XamlX/blob/941dafce490ffd21178f4493a7f0425e9d478e60/src/XamlX/IL/SreTypeSystem.cs#L131

after that, the following test is run that I expected

```c#
    [Fact]
    public void Nestest_Genercic_Class()
    {
        var nestedGenericTypeDefinition = TypeSystem.GetType("TypeSystemTest.Models.Foo`1+Bar`1");
        Assert.NotNull(nestedGenericTypeDefinition);

        Assert.Equal(2, nestedGenericTypeDefinition.GenericParameters.Count);

        var stringType = TypeSystem.GetType("System.String");
        var int32Type = TypeSystem.GetType("System.Int32");
        var boolType = TypeSystem.GetType("System.Boolean");

        var notGenericBarType = nestedGenericTypeDefinition.MakeGenericType(stringType, int32Type);

        var gm = notGenericBarType.Methods[0];

        Assert.NotNull(gm);
        Assert.True(gm.ContainsGenericParameters);
        Assert.Single(gm.GenericParameters);

        var im = gm.MakeGenericMethod([boolType]);
        Assert.NotNull(im);
        Assert.False(im.ContainsGenericParameters);
        Assert.Empty(im.GenericParameters);
    }
```

I think it is not the purpose of this PR to resolve the issue of nested classes.

workgroupengineering committed:
fix: Address review

maxkatz6 commented on tests/TypeSystemTest/BaseTest.cs:
```diff
@@ -0,0 +1,208 @@
+    /// </summary>
+    /// <param name="fullName"></param>
+    /// <returns></returns>
+    private static string GetNormalizzeFullName(string fullName)
```
nit: GetNormalizedFullName

workgroupengineering committed:
fix: nits

workgroupengineering committed:
fix: Conflicts with master

# Conflicts:
#	src/XamlX.IL.Cecil/CecilMethod.cs
#	src/XamlX/IL/SreTypeSystem.cs
